File: Node/Prisma_ORM/src/api/territoryAPI.js
const express = require('express');
const TerritoryService = require('../bll/services/territoryService');
const router = express.Router();

router.get('/', async (req, res) => {
  try {
    const entities = await TerritoryService.getAll();
    res.status(200).json(entities);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

router.get('/:id', async (req, res) => {
  const { id } = req.params;
  try {
    const entity = await TerritoryService.getById(String(id));
    res.status(200).json(entity);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

router.post('/', async (req, res) => {
  const entity = req.body;
  try {
    res.status(200).json(await TerritoryService.create(entity));
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

router.put('/:id', async (req, res) => {
  const entity = req.body;
  const { id } = req.params;
  try {
    res.status(200).json(await TerritoryService.update(String(id), entity));
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

router.delete('/:id', async (req, res) => {
  const { id } = req.params;
  try {
    res.status(200).json(await TerritoryService.delete(String(id)));
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

module.exports = router;

File: Go/GORM_ORM/api/territoryAPI.go
package api

import (
	"net/http"

	"GORM_ORM/bll/services"
	"GORM_ORM/dal/entities"

	"github.com/gin-gonic/gin"
)

type TerritoryAPI struct {
	Service services.TerritoryService
}

func NewTerritoryAPI(s services.TerritoryService) *TerritoryAPI {
	return &TerritoryAPI{Service: s}
}

func (h *TerritoryAPI) GetAll(c *gin.Context) {
	entities, err := h.Service.GetAll()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, entities)
}

func (h *TerritoryAPI) GetByID(c *gin.Context) {
	id := c.Param("id")
	entity, err := h.Service.GetByID(id)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Territory not found"})
		return
	}
	c.JSON(http.StatusOK, entity)
}

func (h *TerritoryAPI) Create(c *gin.Context) {
	var entity entities.Territory
	if err := c.ShouldBindJSON(&entity); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.Service.Create(&entity); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusCreated, entity)
}

func (h *TerritoryAPI) Update(c *gin.Context) {
	id := c.Param("id")
	var entity entities.Territory
	if err := c.ShouldBindJSON(&entity); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	entity.TerritoryID = id
	if err := h.Service.Update(&entity); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, entity)
}

func (h *TerritoryAPI) Delete(c *gin.Context) {
	id := c.Param("id")
	if err := h.Service.Delete(id); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusOK)
}
